### PR TITLE
Implemented contrib.checks and various improvements

### LINKED ISSUE #191: Implement contrib.checks
# Description

Users and potential users are consistently asking for more complex checks. 

We've always supported extensibility in checks, and imagined the type of things that could be implemented for more involved data integrity validation. However, it is clear that the value of goodtables would greatly increase if some such checks were built in to the library, not in the least so that they would eventually be accessible via [goodtables.io](https://goodtables.io).

I recommend that such checks should be shipped as a `contrib` module, in the spirit of [Django Contrib](https://github.com/django/django/tree/master/django/contrib). We could further namespace into groups, for discoverability reasons, but probably not necessary now.

Common requests:

- Check that a column of integers increments
- Declare constraint logic across multiple columns
- Check variation in a column from a mean

# Tasks

- [x] Agree on a design for a contrib module for checks
- [x] Decide on a set of initial checks we can ship based on current needs and feedback
- [x] Implement decided checks

### LINKED ISSUE #198: Finalize custom presets/checks API and checks parameter for validate function
# Overview

For `v1` custom presets and checks are provisional API. As a part of work on `contrib.checks` (#191) we tests all edge cases and finalize those APIs. Also we have make a final decision on how `validate(checks=...)` parameter works. We should take into account that it also should be easy-to-use in CLI and `goodtables.yml`.

# Tasks

- [ ] learn from all edge cases on `contrib.checks`
- [ ] handle feedback on interface provided below
- [ ] discuss `validate(checks)`
- [ ] make changes if needed
- [ ] improve documentation/tutorials

---

From @pwalsh:

Some issues I encountered using goodtables v1 pre.

1. Checks: There is no documentation, and the semantics are not clear. Example: Does `column` refer to a cell in a row, or a column of cells? It must be the former as we are streaming rows, but it is confusing because a column is not a cell.
2. Register: The `check` (and `preset`) decorators are implemented from the `register` module. This suggests that functions decorated as such may get added to a registry of checks and presets. However, we "register" custom checks and presets on the `Inspector` constructor explicitly, So it is not clear to me why we need decorators, when we can just ensure that custom presets and checks implement the signature required to be a `preset` or a `check`. As a user, it feels like extra configuration that I don't need.

### PR DESCRIPTION
**Pull request by roll**
- fixes #191
- fixes #198

---

- implemented following`contrib.checks` (per #191):
  - `blacklisted-value`
  - `deviated-value`
  - `sequential-value`
  - `truncated-value`
  - `custom-constraint`
- renamed `column` to `cell` in checks (per #198)
- improved readability of all builtin checks (more comments etc)
- simplified `checks` argument -> `checks` and `skip_checks` (both lists) (per #198) 
- removed `provisional API` mark from custom checks/presets
- new round of readme improvements to compete `dataproofer`

### CODE REVIEW

**Comment by roll** [comment]
@pwalsh 
I still need to integrate tests for it from different branch and improve documentation but it will be great if you will be able to take a high-level look on new checks - https://github.com/frictionlessdata/goodtables-py/tree/implement-contrib-checks#validation-against-advanced-checks (or advise someone to take a look - Brook, Adam, Dan? - if you don't have time on this week)

**Comment by roll** [comment]
@akariv 
Could you please review? 

There is some noise in the diff because of renaming `columns` to `cells` (#198) and adding comments to existent checks but it's isolated in a few commits. It still passes the same tests so all changes in `goodtables.checks` and `goodtables.presets` could be ignored.

**Comment by roll** [comment]
@georgiana-b 
Please review (Adam is off today)

File: goodtables/contrib/checks/blacklisted_value.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                'column-number': None,
+            })
+
+        # Check value
+        value = cell.get('value')
+        if value in self.__blacklist:
+            message = 'Value "%s" in column %s for row %s is blacklisted'
+            return errors.append({
+                'code': 'blacklisted-value',
+                'message': message % (value, cell['number'], row_number),
```
Comment (roll) [inline goodtables/contrib/checks/blacklisted_value.py]: This one in on purpose to be consistent with Data Quality Spec error messages using column indexes.

**Comment by roll** [comment]
@georgiana-b 
Thanks! I've made changes based on your comments.
